### DISCUSSION TITLE
Narwhal X EVM

### DISCUSSION THREAD

**Discussion by b0xtch** [fleek-network/ursa#227] [Ideas]
Given the requirements for our validators to submit aggregated BLS sigs to our L2. Considering this example [NxE](https://github.com/gakonst/narwhal-abci-evm) could be a great fit for this

@daltoncoder

**Comment by daltoncoder** [comment]
Yes for sure. I’m going to write up some things we’ve discussed on this here, and how we can use this sort of flow. Correct me if I have anything wrong or let me know your thoughts

So this flow doesn’t help us communicate with L2 but it helps us hold state in our network and replicate it consistently accross our nodes. This is because every node is creating their own application state based on the consensus certificates.

So using this flow would could keep our rewards/slashing/transactions in network, and just have a bridge contract on both ends to handle deposits/withdraws(zk proof of consensus bridge)

I like this idea and flow a lot. I also think it future proofs the network a bit and makes things planned in the future a bit easier to come to consensus on(thinking compute mainly here)

**Reply by b0xtch** [reply]
Correct, then some type of oracle/bridge can be used for state transition. I also agree on the future proofing when in the future these computations will be done easily on mobile devices. That opens lots of doors with having mobile nodes, browser nodes, and not just to accelerate content, for p2p computation. So this research is paramount and very much necessary

**Reply by daltoncoder** [reply]
I’m thinking this flow completely removes the need for oracle. Just the bridge. As far as the L2 is concerned it only needs to know what tokens have been bridged and what has been bridged back. All state transitions will happen in network and based on what we come to consensus on

**Reply by b0xtch** [reply]
Interesting, that would even be better, would we have to manage these contract states on chain? And then introduce a native token like originally envisioned?

**Reply by daltoncoder** [reply]
So we would have to manage contract state on our chain, if thats what you mean. And we would do it through abci. Basically every node in our network would be building the application state by submitting the transactions that we have come to consensus on with abci, in the order the consensus has decided. While technically it would not be a native token, we would be able to interact with it in our network like it was and be able to mint and burn tokens in our network before they are bridged back.

**Reply by daltoncoder** [reply]
Im going to try and make a chart to explain this better
